extensions/barrage/src/main/java/io/deephaven/extensions/barrage/table/BarrageTable.java

```diff
@@ -449,18 +449,33 @@ public static BarrageTable make(
 
         final BarrageTable table;
 
-        Object isBlinkTable = attributes.getOrDefault(Table.BLINK_TABLE_ATTRIBUTE, false);
-        if (isBlinkTable instanceof Boolean && (Boolean) isBlinkTable) {
+        final Predicate<String> getAttribute = attr -> {
+            final Object value = attributes.getOrDefault(attr, false);
+            return value instanceof Boolean && (Boolean) value;
+        };
+
+        if (getAttribute.test(Table.BLINK_TABLE_ATTRIBUTE)) {
             final LinkedHashMap<String, ColumnSource<?>> finalColumns = makeColumns(columns, writableSources);
             table = new BarrageBlinkTable(
                     registrar, queue, executor, finalColumns, writableSources, attributes, vpCallback);
         } else {
-            final WritableRowRedirection rowRedirection =
-                    new LongColumnSourceWritableRowRedirection(new LongSparseArraySource());
+            final WritableRowRedirection rowRedirection;
+            final boolean isFlat = getAttribute.test(BarrageUtil.TABLE_ATTRIBUTE_IS_FLAT);
+            if (getAttribute.test(Table.APPEND_ONLY_TABLE_ATTRIBUTE)
+                    || getAttribute.test(Table.ADD_ONLY_TABLE_ATTRIBUTE)
+                    || isFlat) {
+                rowRedirection = new LongColumnSourceWritableRowRedirection(new LongSparseArraySource());
+            } else {
+                rowRedirection = WritableRowRedirectionLockFree.FACTORY.createRowRedirection(1024);
+            }
+
             final LinkedHashMap<String, ColumnSource<?>> finalColumns =
                     makeColumns(columns, writableSources, rowRedirection);
             table = new BarrageRedirectedTable(
                     registrar, queue, executor, finalColumns, writableSources, rowRedirection, attributes, vpCallback);
+            if (isFlat) {
```

Comment:
Maybe we should pass this into the constructor, for consistency with `BarrageBlinkTable` in terms of where we `setFlat`?